### PR TITLE
Add AdEMAMix optimizer

### LINKED ISSUE #33392: Support AdEMAMix optimizer
### Feature request

Add support for the AdEMAMix optimizer

Paper: https://arxiv.org/abs/2409.03137
Implementation: https://github.com/nanowell/AdEMAMix-Optimizer-Pytorch

### Motivation

The AdEMAMix optimizer is a new optimizer from Apple that claims to have better performance and slows down forgetting.

### Your contribution

N/A

### PR DESCRIPTION
**Pull request by matthewdouglas**
# What does this PR do?

Adds support in `Trainer` and `TrainingArguments` for the bitsandbytes implementations of the AdEMAMix optimizer as proposed here: https://hf.co/papers/2409.03137

Compared to AdamW, there is an additional "slow" momentum term to keep older gradients relevant. This authors show that this leads to faster convergence and reduces forgetting during training.

Support for this new optimizer was introduced in [bitsandbytes v0.44.0](https://github.com/bitsandbytes-foundation/bitsandbytes/releases/tag/v0.44.0). This implementation balances out the additional memory overhead when compared to Adam by providing 8bit and 32bit versions that support paging.

Fixes #33392

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@SunMarc

### CODE REVIEW

**Review by SunMarc — APPROVED** [review]
Nice PR @matthewdouglas ! LGTM ! Just a nit

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this optimizer and tests!

**Comment by amyeroberts** [comment]
Merging as failing hub tests are unrelated to this PR

**Comment by umarbutler** [comment]
@matthewdouglas @amyeroberts Thanks for creating this, just a suggestion, however, it would be worth setting 'T<sub>alpha</sub> = T<sub>β3</sub> = T = number of training steps' by default, as recommended by the AdEMAMIx paper, page 6:
image

I tried using this code to train a model not realising that those hyperparameters were not being set by default by `bitsandbytes`, which may explain why I was expericing instability: page 29.

image

Good defaults has always been what sets Hugging Face apart 😁

**Comment by amyeroberts** [comment]
Hi @umarbutler, thanks for flagging! That seems reasonable, especially as we already set defaults for e.g. beta3. Would you like to open a PR to add this?

**Comment by shubhamjain0594** [comment]
@amyeroberts I believe this issue on default values for `t_alpha` and `t_beta3` is still not resolved. Can you confirm? As I cant find anything on default values being set in the repository. I can raise a PR to fix this.
